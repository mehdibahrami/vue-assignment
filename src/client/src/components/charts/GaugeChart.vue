<template>
  <!-- Apex Gauge Chart -->
  <apexchart
    type="radialBar"
    :options="chartOptions"
    :series="series"
  ></apexchart>
</template>

<script>
import VueApexCharts from "vue3-apexcharts";

export default {
  name: "GaugeChart",
  components: {
    apexchart: VueApexCharts,
  },
  props: {
    data: {
      type: Number,
      require: true,
    },
    label: {
      type: String,
      required: true,
    },
    colors: {
      type: Array,
      required: true,
    },
  },
  data() {
    return {
      chartOptions: {
        plotOptions: {
          radialBar: {
            startAngle: -90,
            endAngle: 90,
            hollow: {
              size: "65%",
            },
            track: {
              show: true,
              strokeWidth: "20%",
            },
            dataLabels: {
              name: {
                offsetY: 25,
                color: "#6b7280",
                show: true,
              },
              value: {
                offsetY: -10,
                color: "#8a63a3",
                fontSize: "2rem",
                formatter(val) {
                  return `${parseFloat(val)}`;
                },
              },
            },
          },
        },
        chart: {
          sparkline: {
            enabled: true,
          },
        },
        fill: {
          type: "solid",
          colors: this.colors,
        },
        labels: [this.label],
      },
      series: [this.data],
    };
  },
  watch: {
    data: function () {
      // If data is not null change series value
      if (this.data) {
        this.series = [this.data];
      }
    },
  },
};
</script>
