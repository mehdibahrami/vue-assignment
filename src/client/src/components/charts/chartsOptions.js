import { SPEED_LIMIT, LOW_CHARGE, MID_CHARGE } from "@/constants";

const defaultGaugeChart = {
  plotOptions: {
    radialBar: {
      startAngle: -90,
      endAngle: 90,
      hollow: {
        size: "65%",
      },
      track: {
        show: true,
        strokeWidth: "20%",
      },
      dataLabels: {
        name: {
          offsetY: 25,
          color: "#6b7280",
          show: true,
        },
        value: {
          offsetY: -10,
          color: "#8a63a3",
          fontSize: "2rem",
          formatter(val) {
            return `${parseFloat(val)}`;
          },
        },
      },
    },
  },
  chart: {
    sparkline: {
      enabled: true,
    },
  },
};

export const speedGaugeChartOptions = {
  ...defaultGaugeChart,
  fill: {
    colors: [
      function ({ value }) {
        if (value <= SPEED_LIMIT) {
          return "#7ca363";
        } else {
          return "#AD6D6D";
        }
      },
    ],
    type: "solid",
  },
  labels: ["km/h"],
};

export const socGaugeChartOptions = {
  ...defaultGaugeChart,
  fill: {
    colors: [
      function ({ value }) {
        if (value <= LOW_CHARGE) {
          return "#AD6D6D";
        } else if (value > LOW_CHARGE && value <= MID_CHARGE) {
          return "#C38540";
        } else {
          return "#7ca363";
        }
      },
    ],
    type: "solid",
  },
  labels: ["%"],
};

const defaultLineChart = {
  dataLabels: {
    enabled: false,
  },
  stroke: {
    curve: "smooth",
  },
};

export const speedLineChartOptions = {
  ...defaultLineChart,
  chart: {
    id: "realtime",
    type: "line",
    zoom: {
      enabled: false,
    },
    toolbar: {
      show: false,
    },
    animations: {
      enabled: true,
      speed: 100,
      easing: "linear",
      dynamicAnimation: {
        enabled: true,
        speed: 500,
      },
    },
  },
  colors: ["#8a63a3"],
  xaxis: {
    tickAmount: 5,
    range: 100000,
    labels: {
      formatter: function (val, timestamp) {
        return new Date(timestamp).toLocaleTimeString([], {
          hour12: false,
        });
      },
      offsetX: -15,
    },
  },
  yaxis: {
    min: 0,
    max: 100,
    tickAmount: 5,
    opposite: true,
    labels: {
      offsetX: -10,
    },
  },
  tooltip: {
    x: {
      show: true,
      formatter: function (val) {
        return new Date(val).toLocaleTimeString([], {
          hour12: false,
        });
      },
    },
    y: {
      formatter: function (value) {
        return value + " km/h";
      },
      title: "Speed",
    },
  },
};
