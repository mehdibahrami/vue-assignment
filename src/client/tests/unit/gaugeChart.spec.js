import { mount } from "@vue/test-utils";
import GaugeChart from "@/components/charts/GaugeChart.vue";
import { speedGaugeChartOptions } from "@/components/charts/chartsOptions.js";

describe("GaugeChart", () => {
  let wrapper;

  it("renders correctly", () => {
    // shallow mounting the main component
    wrapper = mount(GaugeChart, {
      props: {
        options: speedGaugeChartOptions,
        data: 4,
      },
    });

    expect(wrapper.findComponent({ name: "apexchart" }).exists()).toBeTruthy();
    expect(wrapper.find(".vue-apexcharts").exists()).toBeTruthy();
    expect(wrapper.element).toMatchSnapshot();
  });
});
