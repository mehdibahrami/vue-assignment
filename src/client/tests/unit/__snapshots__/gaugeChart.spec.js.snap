// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GaugeChart renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <!-- Apex Gauge Chart -->
  <div
    class="vue-apexcharts"
  />
  
</div>
`;
